### PR TITLE
Add numeric separator lexing

### LINKED ISSUE #994: Numeric separators are not supported
**Describe the bug**
Numeric separators are not supported.

**To Reproduce**
Steps to reproduce the issue, or JavaScript code that causes this failure.

```javascript
var a = 1_0;
a;
```

**Expected behavior**

Running this code, `a` should be set to `10` and printed, but a syntax error is generated instead. The expected behaviour can be found in the [ECMAScript specification][spec].

[spec]: https://tc39.es/ecma262/#sec-literals-numeric-literals

**Build environment (please complete the following information):**

- OS: Windows 10
- Version: 0.10
- Target triple: x86_64-pc-windows-msvc
- Rustc version: rustc 1.48.0 (7eac88abb 2020-11-16)

**Additional context**
None

### PR DESCRIPTION
**Pull request by tofpie**
This Pull Request fixes #994 

It changes the following:

- Lexing of numbers

### CODE REVIEW

File: boa/src/syntax/lexer/number.rs [hunk 1]
```diff
@@ -110,11 +110,52 @@ where
...
+    buf: &mut Vec<u8>,
+    cursor: &mut Cursor<R>,
+    kind: &NumericKind,
+    separator_allowed: bool,
+) -> Result<(), Error>
+where
+    R: Read,
+{
+    let mut prev_is_underscore = false;
+    let mut pos = cursor.pos();
```
Comment (HalidOdat) [inline boa/src/syntax/lexer/number.rs:128]: Do we need the `pos` variable, cant we call `cursor.pos()` where we need the position?

File: boa/src/syntax/lexer/number.rs [hunk 2]
```diff
@@ -253,34 +294,54 @@ impl<R> Tokenizer<R> for NumberLiteral {
...
             Some(b'.') => {
                 if kind.base() == 10 {
-                    // Only base 10 numbers can have a decimal seperator.
+                    // Only base 10 numbers can have a decimal separator.
                     // Number literal lexing finished if a . is found for a number in a different base.
-
-                    cursor.next_byte()?.expect(". token vanished");
-                    buf.push(b'.'); // Consume the .
+                    if self.init != b'.' {
+                        cursor.next_byte()?.expect(". token vanished");
```
Comment (HalidOdat) [inline boa/src/syntax/lexer/number.rs:330]:
```suggestion
                        cursor.next_byte()?.expect("'.' token vanished");
```
It's clearer to put the token in `'` (single quote) so we don't confuse it with punctuation from the text.

File: boa/src/syntax/lexer/number.rs [hunk 3]
```diff
@@ -253,34 +294,54 @@ impl<R> Tokenizer<R> for NumberLiteral {
...
+                        buf.push(b'.'); // Consume the .
+                    }
                     kind = NumericKind::Rational;
 
-                    // Consume digits until a non-digit character is encountered or all the characters are consumed.
-                    cursor.take_while_ascii_pred(&mut buf, &|c: char| c.is_digit(kind.base()))?;
+                    if cursor.peek()? == Some(b'_') {
+                        return Err(Error::syntax(
+                            "numeric separator not allowed after .",
```
Comment (HalidOdat) [inline boa/src/syntax/lexer/number.rs:337]:
```suggestion
                            "numeric separator not allowed after '.'",
```
same here.

File: boa/src/syntax/lexer/number.rs [hunk 4]
```diff
@@ -253,34 +294,54 @@ impl<R> Tokenizer<R> for NumberLiteral {
...
-        match cursor.peek()? {
+        match next {
             Some(b'n') => {
                 // DecimalBigIntegerLiteral
                 // Lexing finished.
-
                 // Consume the n
+                if legacy_octal {
+                    return Err(Error::syntax(
+                        "n suffix not allowed in octal representation",
```
Comment (HalidOdat) [inline boa/src/syntax/lexer/number.rs:317]:
```suggestion
                        "'n' suffix not allowed in octal representation",
```

**Review by HalidOdat — COMMENTED** [review]
Looks good!

Check my comments :)

**Comment by HalidOdat** [comment]
| Test result | master count | PR count | difference |
| :---------: | :----------: | :------: | :--------: |
| Total | 78,493 | 78,493 | 0 |
| Passed | 19,596 | 19,735 | **+139** |
| Ignored | 15,585 | 15,585 | 0 |
| Failed | 43,312 | 43,173 | **-139** |
| Panics | 447 | 447 | 0 |
| Conformance | 24.97 | 25.14 | **+0.18%** |

File: boa/src/syntax/lexer/number.rs (hunk 1 above)
Comment (tofpie) [reply boa/src/syntax/lexer/number.rs:128]: This is to have the correct position when the separator is at the end of the number. If we take `cursor.pos()`, then the position is after the separator, on not on the separator.

**Review by RageKnify — APPROVED** [review]
Looks good to me, but I'd preffer if one of the "parser people" had a look since I still don't fully grasp it.

**Review by Lan2u — APPROVED** [review]
Looks good to me 👍

**Pull request merged**